src/Features/Core/Portable/FileHeaders/AbstractFileHeaderHelper.cs

```diff
@@ -0,0 +1,152 @@
+        private int EndOfLineTriviaKind { get; }
+
+        protected abstract string GetTextContextOfComment(SyntaxTrivia commentTrivia);
+
+        public FileHeader ParseFileHeader(SyntaxNode root)
+        {
+            var firstToken = root.GetFirstToken(includeZeroWidth: true);
+            var firstNonWhitespaceTrivia = IndexOfFirstNonWhitespaceTrivia(firstToken.LeadingTrivia, true);
+
+            if (firstNonWhitespaceTrivia == -1)
+            {
+                return FileHeader.MissingFileHeader(0);
+            }
+
+            var sb = StringBuilderPool.Allocate();
+            var endOfLineCount = 0;
+            var done = false;
+            var missingHeaderOffset = 0;
+            var fileHeaderStart = int.MaxValue;
+            var fileHeaderEnd = int.MinValue;
+
+            for (var i = firstNonWhitespaceTrivia; !done && (i < firstToken.LeadingTrivia.Count); i++)
+            {
+                var trivia = firstToken.LeadingTrivia[i];
+
+                if (trivia.RawKind == WhitespaceTriviaKind)
+                {
+                    endOfLineCount = 0;
+                }
+                else if (trivia.RawKind == SingleLineCommentTriviaKind)
+                {
+                    endOfLineCount = 0;
+
+                    var commentText = GetTextContextOfComment(trivia);
+
+                    fileHeaderStart = Math.Min(trivia.FullSpan.Start, fileHeaderStart);
+                    fileHeaderEnd = trivia.FullSpan.End;
+
+                    sb.AppendLine(commentText.Trim());
+                }
+                else if (trivia.RawKind == MultiLineCommentTriviaKind)
+                {
+                    // only process a MultiLineCommentTrivia if no SingleLineCommentTrivia have been processed
+                    if (sb.Length == 0)
+                    {
+                        var commentText = GetTextContextOfComment(trivia);
+                        var triviaStringParts = commentText.Trim().Replace("\r\n", "\n").Split('\n');
+
+                        foreach (var part in triviaStringParts)
+                        {
+                            var trimmedPart = part.TrimStart(' ', '*');
+                            sb.AppendLine(trimmedPart);
+                        }
+
+                        fileHeaderStart = trivia.FullSpan.Start;
+                        fileHeaderEnd = trivia.FullSpan.End;
+                    }
+
+                    done = true;
+                }
+                else if (trivia.RawKind == EndOfLineTriviaKind)
+                {
+                    endOfLineCount++;
+                    done = endOfLineCount > 1;
+                }
+                else
+                {
+                    if (trivia.IsDirective)
+                    {
+                        missingHeaderOffset = trivia.FullSpan.End;
+                    }
+
+                    done = (fileHeaderStart < fileHeaderEnd) || !trivia.IsDirective;
+                }
+            }
```

Comment:
i found this basically entirely to complex to understand.  consider instead a helper that first collects the requisite trivia, then in a later step takes that trivia to build the header string.

Reply:
This one might not be perfect, but I feel comfortable leaving this one for now with the test coverage.